### PR TITLE
test: test initramfs

### PR DESCRIPTION
**Pull request by WoodenMaiden**

### CODE REVIEW

**Comment by lucido-simon** [comment]
Hi! Thanks for the PR. Can you squash the two commits together please ?

**Review by alexis-langlet — CHANGES_REQUESTED** [review]
Nice work
The only real issue of all this is that you don't use mocks as mocks but as different implementations

File: initramfs/src/image.rs [hunk 1]
```diff
@@ -175,3 +195,115 @@ impl Image {
...
+            Ok(Self {
+                path: path.to_string(),
+            })
+        }
+
+        fn open(path: &str) -> anyhow::Result<Self>
+        where
+            Self: std::marker::Sized,
+        {
+            MockFileHandler::create(path)
```
Comment (alexis-langlet) [inline initramfs/src/image.rs:238]: There could be some problems if someone needs to change create() behaviour 

```suggestion
            Ok(Self {
                path: path.to_string(),
            })
```

File: initramfs/src/image.rs [hunk 2]
```diff
@@ -175,3 +195,115 @@ impl Image {
...
+        impl Write for MockFileHandler {
+            fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+                unsafe {
+                    if self.path.contains(".img") && FILE.len() < 2 {
+                        FILE.extend_from_slice(buf)
+                    }                 
+                };
+
+                std::result::Result::Ok(buf.len())
+            }
```
Comment (alexis-langlet) [inline initramfs/src/image.rs:282]: This part seems a bit weird to me, it would be nice having Simon's POV on this

**Review by alexis-langlet — DISMISSED** [review]
Really nice job, thank you

File: initramfs/src/image.rs [hunk 3]
```diff
@@ -12,6 +12,26 @@ use log::debug;
...
 
+/// Trait to abstract File reading
+pub trait FileHandler: Write + Read {
+    fn create(path: &str) -> Result<Self>
+    where
+        Self: std::marker::Sized;
+    fn open(path: &str) -> Result<Self>
+    where
+        Self: std::marker::Sized;
+}
```
Comment (lucido-simon) [inline initramfs/src/image.rs:23]:
```suggestion
pub trait FileHandler: Write + Read + Sized {
    fn create(path: &str) -> Result<Self>;
    fn open(path: &str) -> Result<Self>;
}
```

Looks cleaner imo

File: initramfs/src/image.rs [hunk 4]
```diff
@@ -61,16 +86,17 @@ impl Image {
...
         init_path: &str,
         agent_path: &str,
         agent_config_path: &str,
-    ) -> Result<()> {
+    ) -> Result<Handler> {
         // Write the cpio to disk
         let file_name = format!("initramfs-{}-{}.img", self.name.replace('/', "-"), self.tag);
         let archive = Encoder::new(
-            File::create(file_name).map_err(|e| anyhow!(e).context("Failed to create file"))?,
+            <Handler>::create(&file_name)
```
Comment (lucido-simon) [inline initramfs/src/image.rs:98]:
```suggestion
            Handler::create(&file_name)
```

File: initramfs/src/image.rs [hunk 5]
```diff
@@ -126,11 +152,11 @@ impl Image {
             }
         }
 
-        let mut init_file =
-            File::open(init_path).map_err(|e| anyhow!(e).context("Failed to open init file"))?;
-        let mut agent_file =
-            File::open(agent_path).map_err(|e| anyhow!(e).context("Failed to open init file"))?;
-        let mut agent_config_file = File::open(agent_config_path)
+        let mut init_file = <Handler>::open(init_path)
```
Comment (lucido-simon) [inline initramfs/src/image.rs:155]:
```suggestion
        let mut init_file = Handler::open(init_path)
```

Probably need to run `rustfmt` after that, it's probably not going to format it the same way.

File: initramfs/src/image.rs [hunk 6]
```diff
@@ -126,11 +152,11 @@ impl Image {
...
+            .map_err(|e| anyhow!(e).context("Failed to open init file"))?;
+        let mut agent_file = <Handler>::open(agent_path)
```
Comment (lucido-simon) [inline initramfs/src/image.rs:157]:
```suggestion
        let mut agent_file = Handler::open(agent_path)
```

Same

File: initramfs/src/image.rs [hunk 7]
```diff
@@ -126,11 +152,11 @@ impl Image {
...
+            .map_err(|e| anyhow!(e).context("Failed to open init file"))?;
+        let mut agent_config_file = <Handler>::open(agent_config_path)
```
Comment (lucido-simon) [inline initramfs/src/image.rs:159]:
```suggestion
        let mut agent_config_file = Handler::open(agent_config_path)
```
Same

**Review by lucido-simon — CHANGES_REQUESTED** [review]
LGTM, except a few formatting nits. gj !

**Comment by lucido-simon** [comment]
I've taken the liberty of making a small change: renaming `blobSum` of the `LayerMetadata` struct to `blob_sum`. This allows removing the `#[allow(non_snake_case)]`, by replacing it by `#[serde(rename_all = "camelCase")]` in `image.rs`.

Other than that, PR LGTM.

**Review by lucido-simon — APPROVED** [review]

**Pull request merged**